FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1892 @@
+#include "FEXCore/Utils/LogManager.h"
+#include "Interface/IR/IR.h"
+#include "Interface/IR/IREmitter.h"
+#include "Interface/IR/PassManager.h"
+#include <FEXCore/IR/IR.h>
+#include <FEXCore/Utils/Profiler.h>
+#include <FEXCore/fextl/deque.h>
+
+#include <array>
+#include <cstddef>
+#include <cstdint>
+#include <stdint.h>
+
+namespace FEXCore::IR {
+
+// FixedSizeStack is a model of the x87 Stack where each element in this
+// fixed size stack lives at an offset from top. The top of the stack is at
+// index 0.
+template<typename T, size_t MaxSize>
+class FixedSizeStack {
+public:
+  enum class Slot { UNUSED, INVALID, VALID };
+  FixedSizeStack() {
+    buffer.resize(MaxSize, {Slot::UNUSED, T()});
+  }
+
+  void push(const T& value) {
+    rotate();
+    buffer.front() = {Slot::VALID, value};
+  }
```

Comment:
This all seems very contrived, seemingly with the intent of improving performance, but `std::deque` is a heavyweight container that interferes with data caching and branch prediction.

Have you considered just using an `std::array` and reversing the order elements are stored internally? That would make the implementation a lot more straightforward while having little (if any) impact on performance: Indexing in `top` will be somewhat faster, and `rotate` slightly slower (not by much since we're talking about tiny amounts of data though).

Reply:
My initial impl used an array but for some reason that is escaping me now, I then switched to a deque. I will need to think a bit more about why I switched and the considerations here. There are quite a few corner cases that made using a deque easier.